### DISCUSSION TITLE
Cannot resolve method .newJavaKeystoreBuilder() when using Aws-iot-device-sdk-java-v2 AwsIotMqttConnectionBuilder

### DISCUSSION THREAD

**Discussion by Devoamos** [aws/aws-iot-device-sdk-java-v2#347] [Q&A] [answered]
Hi,

I am following this [sample code](https://github.com/aws/aws-iot-device-sdk-java-v2/blob/50b756ab43276070ce62de1e4001c7c9aa347724/samples/Utils/CommandLineUtils/utils/commandlineutils/CommandLineUtils.java#L324) to create the JavaKeyStore connection in order to migrate from SDK v1 -> v2. However, when importing `AwsIotMqttConnectionBuilder` that method does not seem to exist in the root files. Is there anything I am doing wrong or have I missed something? 

![Screenshot 2023-01-11 at 11 51 38](https://user-images.githubusercontent.com/69257041/211799529-5daa3bf2-dca0-494d-abc6-5e74e3182f6e.png)

I am using the following library with Java 11 and Spring 2.7.3

`<dependency>
  <groupId>software.amazon.awssdk.iotdevicesdk</groupId>
  <artifactId>aws-iot-device-sdk</artifactId>
  <version>1.11.0</version>
</dependency>`

**Comment by bretambrose** [comment]
That was added in 1.10.6

https://github.com/aws/aws-iot-device-sdk-java-v2/releases/tag/v1.10.6

**Reply by Devoamos** [reply]
Thanks.

However, I am using version 1.11.0, would it not also be present in there?

**Reply by TwistedTwigleg** [reply]
Is your dependency set to the following:
`<dependency> <groupId>software.amazon.awssdk.iotdevicesdk</groupId> <artifactId>aws-iot-device-sdk</artifactId> <version>1.11.0</version> </dependency>`? In your original post you mentioned it was set to `1.10.5` which does not have Java Keystore support, which I think is where the confusion is.
In your image posted though, it does appear you are looking at version `1.11.0`, but if your `pom.xml` file is using version `1.10.5` then it will not work until you use version `1.10.6` or later (including version `1.11.0` - it should have full support).

In the source code, the function to use a Java keystore to connect should be around line `191` in `AwsIotMqttConnectionBuilder`. Can you please check to see if the function is present? The function should be called `newJavaKeystoreBuilder`.

![Screenshot 2023-01-13 at 09-19-18 aws_aws-iot-device-sdk-java-v2 Next generation AWS IoT Client SDK for Java using the AWS Common Runtime](https://user-images.githubusercontent.com/25082678/212341873-0ccb3772-f17b-4e50-b340-c877fa22bc1f.png)

**Reply by Devoamos** [reply]
Sorry, the correct version is 1.11.0 I have amended that now. 

This is unusual so I'm not sure where it's going wrong. 

![Screenshot 2023-01-16 at 09 29 58](https://user-images.githubusercontent.com/69257041/212645863-756f3f45-8f40-4082-b103-4c29f7ca10a8.png)


![Screenshot 2023-01-16 at 09 39 39](https://user-images.githubusercontent.com/69257041/212646481-1b8329ab-319b-4c44-a17f-1265db1091f2.png)

To build on this question, we have decided to go with the `.newMtlsBuilderFromPath()` but can see the below error on startup.

```
{"time":"2023-01-16T09:50:05.086Z","level":"INFO","thread":"main","method":"logMessage","msg":"__Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.","logger":"ConditionEvaluationReportLoggingListener","exception":"","logger_name":"org.springframework.boot.autoconfigure.logging.ConditionEvaluationReportLoggingListener"}
{"time":"2023-01-16T09:50:05.299Z","level":"ERROR","thread":"main","method":"reportFailure","msg":"Application run failed","logger":"SpringApplication","exception":"org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'messagePublisher' defined in URL [jar:file:/app/app.jar!/BOOT-INF/classes!/uk/co/argos/convenience/ccutils/iot/MessagePublisher.class]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [uk.co.argos.convenience.ccutils.iot.MessagePublisher]: Constructor threw exception; nested exception is java.lang.UnsatisfiedLinkError: /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so: Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so)\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:315)\n\tat org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:296)\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [uk.co.argos.convenience.ccutils.iot.MessagePublisher]: Constructor threw exception; nested exception is java.lang.UnsatisfiedLinkError: /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so: Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so)\n\tat org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224)\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117)\nCaused by: java.lang.UnsatisfiedLinkError: /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so: Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /tmp/AWSCRT_3245581712843744318libaws-crt-jni.so)\n\tat java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)\n\tat java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)\n","logger_name":"org.springframework.boot.SpringApplication"}
```

Does anyone know what the root cause could be?

Thanks.

**Reply by jmklix** [reply]
Can you provide reproduction steps for getting the same error you mentioned above?

**Reply by jmklix** [reply, marked as answer]
Please open a new discussion with detailed reproduction steps if you are still running into this issue
